File: src/dictation/time_map.py
"""Sample arithmetic and interval algebra (plan sections 6 and 7).

Two rules run through this module:

* Intervals are half-open integer sample ranges, so concatenation is exact and
  a sample belongs to exactly one side of a boundary.
* Every rounding on a *removal* interval rounds outward, and every rounding on
  a *retained* interval rounds inward.  A rounding error must never leave part
  of a removed word in the output.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Iterable, Sequence

from dictation.types import CANONICAL_SAMPLE_RATE, SampleInterval, Word


def ms_to_samples(ms: float, sample_rate: int = CANONICAL_SAMPLE_RATE) -> int:
    """Round-half-up conversion for durations such as padding."""
    return int(round(ms * sample_rate / 1000.0))


def samples_to_ms(samples: int, sample_rate: int = CANONICAL_SAMPLE_RATE) -> float:
    return samples * 1000.0 / sample_rate


def normalize(intervals: Iterable[SampleInterval]) -> tuple[SampleInterval, ...]:
    """Sort, drop empties and merge overlapping or abutting intervals."""
    ordered = sorted((i for i in intervals if not i.is_empty), key=lambda i: (i.start, i.end))
    merged: list[SampleInterval] = []
    for interval in ordered:
        if merged and interval.start <= merged[-1].end:
            previous = merged[-1]
            if interval.end > previous.end:
                merged[-1] = SampleInterval(previous.start, interval.end)
        else:
            merged.append(interval)
    return tuple(merged)


def pad(
    interval: SampleInterval,
    padding_samples: int,
    bound: SampleInterval,
) -> SampleInterval:
    """Grow a removal interval outward, clipped to the available audio."""
    if padding_samples < 0:
        raise ValueError("padding must be non-negative")
    start = max(bound.start, interval.start - padding_samples)
    end = min(bound.end, interval.end + padding_samples)
    return SampleInterval(start, max(start, end))


def pad_all(
    intervals: Iterable[SampleInterval],
    padding_samples: int,
    bound: SampleInterval,
) -> tuple[SampleInterval, ...]:
    return normalize(pad(i, padding_samples, bound) for i in intervals)


def total_length(intervals: Iterable[SampleInterval]) -> int:
    return sum(i.length for i in intervals)


def complement(
    removed: Sequence[SampleInterval],
    bound: SampleInterval,
) -> tuple[SampleInterval, ...]:
    """Retained intervals: ``bound`` minus ``removed`` (plan section 7.8)."""
    retained: list[SampleInterval] = []
    cursor = bound.start
    for interval in normalize(removed):
        clipped = interval.clamp(bound)
        if clipped.is_empty:
            continue
        if clipped.start > cursor:
            retained.append(SampleInterval(cursor, clipped.start))
        cursor = max(cursor, clipped.end)
    if cursor < bound.end:
        retained.append(SampleInterval(cursor, bound.end))
    return tuple(retained)


def intersects_any(interval: SampleInterval, intervals: Iterable[SampleInterval]) -> bool:
    return any(interval.intersects(other) for other in intervals)


def contains_interval(outer: Sequence[SampleInterval], inner: SampleInterval) -> bool:
    """True when ``inner`` lies wholly inside one of ``outer``."""
    return any(o.start <= inner.start and inner.end <= o.end for o in normalize(outer))


@dataclass(frozen=True, slots=True)
class DestinationMap:
    """Maps retained source samples onto the concatenated export timeline.

    For retained source interval ``[a, b)`` following retained intervals of
    total length ``L``, the destination interval is ``[L, L + b - a)`` and a
    retained sample ``t`` maps to ``L + t - a`` (plan section 7).
    """

    retained: tuple[SampleInterval, ...]
    offsets: tuple[int, ...]

    @classmethod
    def build(cls, retained: Sequence[SampleInterval]) -> DestinationMap:
        normalized = normalize(retained)
        offsets: list[int] = []
        running = 0
        for interval in normalized:
            offsets.append(running)
            running += interval.length
        return cls(tuple(normalized), tuple(offsets))

    @property
    def total_samples(self) -> int:
        return total_length(self.retained)

    def destination_of(self, index: int) -> SampleInterval:
        """Destination interval of the ``index``-th retained interval."""
        offset = self.offsets[index]
        return SampleInterval(offset, offset + self.retained[index].length)

    def map_sample(self, sample: int) -> int:
        """Destination position of a retained source sample.

        Raises ``KeyError`` for a removed sample: a removed position has no
        destination, and silently clamping it would fabricate alignment.
        """
        for index, interval in enumerate(self.retained):
            if interval.contains(sample):
                return self.offsets[index] + sample - interval.start
        raise KeyError(f"sample {sample} is not in retained audio")

    def map_word(self, word: Word) -> SampleInterval:
        """Destination interval of a word that lies wholly inside one clip."""
        for index, interval in enumerate(self.retained):
            if interval.start <= word.start_sample and word.end_sample <= interval.end:
                offset = self.offsets[index] - interval.start
                return SampleInterval(word.start_sample + offset, word.end_sample + offset)
        raise KeyError(f"word {word.id} crosses or misses a retained interval")

    def boundary_manifest(self) -> tuple[dict[str, int], ...]:
        """Per-clip boundary record kept with a spliced export (plan 7.8).

        Source positions stay local: they are never uploaded, because a removal
        map would explain what was deleted (plan section 7, last paragraph).
        """
        return tuple(
            {
                "clip_index": index,
                "source_start": interval.start,
                "source_end": interval.end,
                "destination_start": self.offsets[index],
                "destination_end": self.offsets[index] + interval.length,
            }
            for index, interval in enumerate(self.retained)
        )


@dataclass(frozen=True, slots=True)
class ResampleMap:
    """Mapping between the stored signal and canonical 16 kHz sample indices.

    Capture may arrive at 44.1 or 48 kHz.  The canonical stream is what the ASR
    sees and what spans refer to, but cuts must apply to the exact stored
    signal, so the mapping is kept rather than recomputed from a ratio at use
    time (plan section 6).
    """

    source_rate: int
    canonical_rate: int = CANONICAL_SAMPLE_RATE

    def __post_init__(self) -> None:
        if self.source_rate <= 0 or self.canonical_rate <= 0:
            raise ValueError("sample rates must be positive")

    @property
    def ratio(self) -> float:
        return self.source_rate / self.canonical_rate

    def to_source_sample(self, canonical_sample: int) -> int:
        return int(math.floor(canonical_sample * self.ratio))

    def to_canonical_sample(self, source_sample: int) -> int:
        return int(math.floor(source_sample / self.ratio))

    def removal_to_source(self, interval: SampleInterval) -> SampleInterval:
        """Round a removal interval outward in the source timeline."""
        start = int(math.floor(interval.start * self.ratio))
        end = int(math.ceil(interval.end * self.ratio))
        return SampleInterval(max(0, start), max(0, end))

    def retained_to_source(self, interval: SampleInterval) -> SampleInterval:
        """Round a retained interval inward in the source timeline."""
        start = int(math.ceil(interval.start * self.ratio))
        end = int(math.floor(interval.end * self.ratio))
        return SampleInterval(start, max(start, end))


class AlignmentProblem(str):
    """Reason code describing why an alignment is unusable."""


def validate_alignment(
    words: Sequence[Word],
    bound: SampleInterval,
    *,
    min_confidence: float,
) -> tuple[str, ...]:
    """Return reason codes for anything that makes a cut unprovable.

    Plan section 7.7: missing, overlapping, non-monotonic, low-confidence or
    transcript-inconsistent timings discard the utterance or the session.  This
    function only reports; the caller decides the scope of the discard.
    """
    problems: list[str] = []
    previous_end = bound.start
    for word in words:
        if not word.text.strip():
            problems.append("empty_word_text")
        if word.start_sample == word.end_sample:
            problems.append("zero_length_word")
        if word.start_sample < previous_end:
            problems.append("overlapping_or_nonmonotonic")
        if word.start_sample < bound.start or word.end_sample > bound.end:
            problems.append("word_outside_audio")
        if word.confidence < min_confidence:
            problems.append("low_confidence_word")
        if word.timing_unreliable:
            problems.append("unreliable_timing")
        previous_end = word.end_sample
    return tuple(dict.fromkeys(problems))

File: crates/dictation-core/src/time_map.rs
//! Half-open sample interval arithmetic used by privacy filtering.
//!
//! Removal intervals always round outward while retained intervals round
//! inward. This ensures rounding cannot leave a fragment of removed audio in a
//! retained clip.

use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SampleInterval {
    start: u64,
    end: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct InvertedInterval {
    pub start: u64,
    pub end: u64,
}

impl fmt::Display for InvertedInterval {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "inverted interval: [{}, {})",
            self.start, self.end
        )
    }
}

impl std::error::Error for InvertedInterval {}

impl SampleInterval {
    #[must_use]
    pub const fn from_zero(end: u64) -> Self {
        Self { start: 0, end }
    }

    /// Construct a half-open interval `[start, end)`.
    ///
    /// # Errors
    ///
    /// Returns [`InvertedInterval`] when `end` precedes `start`.
    pub fn new(start: u64, end: u64) -> Result<Self, InvertedInterval> {
        if end < start {
            return Err(InvertedInterval { start, end });
        }
        Ok(Self { start, end })
    }

    #[must_use]
    pub const fn start(self) -> u64 {
        self.start
    }

    #[must_use]
    pub const fn end(self) -> u64 {
        self.end
    }

    #[must_use]
    pub const fn len(self) -> u64 {
        self.end - self.start
    }

    #[must_use]
    pub const fn is_empty(self) -> bool {
        self.start == self.end
    }

    #[must_use]
    pub const fn contains(self, sample: u64) -> bool {
        self.start <= sample && sample < self.end
    }

    #[must_use]
    pub const fn intersects(self, other: Self) -> bool {
        self.start < other.end && other.start < self.end
    }

    #[must_use]
    pub const fn touches(self, other: Self) -> bool {
        self.start <= other.end && other.start <= self.end
    }

    #[must_use]
    pub fn clamp(self, bound: Self) -> Self {
        let start = self.start.clamp(bound.start, bound.end);
        let end = self.end.clamp(bound.start, bound.end).max(start);
        Self { start, end }
    }
}

#[must_use]
pub fn milliseconds_to_samples(milliseconds: u64, sample_rate: u32) -> u64 {
    let numerator = u128::from(milliseconds)
        .saturating_mul(u128::from(sample_rate))
        .saturating_add(500);
    u64::try_from(numerator / 1_000).unwrap_or(u64::MAX)
}

#[must_use]
#[allow(clippy::cast_precision_loss)]
pub fn samples_to_milliseconds(samples: u64, sample_rate: u32) -> f64 {
    samples as f64 * 1_000.0 / f64::from(sample_rate)
}

#[must_use]
pub fn normalize(intervals: impl IntoIterator<Item = SampleInterval>) -> Vec<SampleInterval> {
    let mut ordered: Vec<_> = intervals
        .into_iter()
        .filter(|interval| !interval.is_empty())
        .collect();
    ordered.sort_unstable();

    let mut merged: Vec<SampleInterval> = Vec::with_capacity(ordered.len());
    for interval in ordered {
        if let Some(previous) = merged.last_mut() {
            if interval.start <= previous.end {
                previous.end = previous.end.max(interval.end);
                continue;
            }
        }
        merged.push(interval);
    }
    merged
}

#[must_use]
pub fn pad(
    interval: SampleInterval,
    padding_samples: u64,
    bound: SampleInterval,
) -> SampleInterval {
    let start = interval
        .start
        .saturating_sub(padding_samples)
        .max(bound.start);
    let end = interval
        .end
        .saturating_add(padding_samples)
        .min(bound.end)
        .max(start);
    SampleInterval { start, end }
}

#[must_use]
pub fn pad_all(
    intervals: impl IntoIterator<Item = SampleInterval>,
    padding_samples: u64,
    bound: SampleInterval,
) -> Vec<SampleInterval> {
    normalize(
        intervals
            .into_iter()
            .map(|interval| pad(interval, padding_samples, bound)),
    )
}

#[must_use]
pub fn complement(removed: &[SampleInterval], bound: SampleInterval) -> Vec<SampleInterval> {
    let mut retained = Vec::new();
    let mut cursor = bound.start;
    for interval in normalize(removed.iter().copied()) {
        let clipped = interval.clamp(bound);
        if clipped.is_empty() {
            continue;
        }
        if clipped.start > cursor {
            retained.push(SampleInterval {
                start: cursor,
                end: clipped.start,
            });
        }
        cursor = cursor.max(clipped.end);
    }
    if cursor < bound.end {
        retained.push(SampleInterval {
            start: cursor,
            end: bound.end,
        });
    }
    retained
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DestinationMap {
    retained: Vec<SampleInterval>,
    offsets: Vec<u64>,
}

impl DestinationMap {
    #[must_use]
    pub fn build(retained: impl IntoIterator<Item = SampleInterval>) -> Self {
        let retained = normalize(retained);
        let mut running = 0_u64;
        let offsets = retained
            .iter()
            .map(|interval| {
                let current = running;
                running = running.saturating_add(interval.len());
                current
            })
            .collect();
        Self { retained, offsets }
    }

    #[must_use]
    pub fn retained(&self) -> &[SampleInterval] {
        &self.retained
    }

    #[must_use]
    pub fn total_samples(&self) -> u64 {
        self.retained.iter().map(|interval| interval.len()).sum()
    }

    #[must_use]
    pub fn destination_of(&self, index: usize) -> Option<SampleInterval> {
        let source = *self.retained.get(index)?;
        let offset = *self.offsets.get(index)?;
        Some(SampleInterval {
            start: offset,
            end: offset + source.len(),
        })
    }

    #[must_use]
    pub fn map_sample(&self, sample: u64) -> Option<u64> {
        self.retained
            .iter()
            .zip(&self.offsets)
            .find(|(interval, _)| interval.contains(sample))
            .map(|(interval, offset)| offset + sample - interval.start)
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ResampleMap {
    source_rate: u32,
    canonical_rate: u32,
}

impl ResampleMap {
    /// Construct a mapping between canonical and stored sample rates.
    ///
    /// # Errors
    ///
    /// Returns an error if either sample rate is zero.
    pub fn new(source_rate: u32, canonical_rate: u32) -> Result<Self, &'static str> {
        if source_rate == 0 || canonical_rate == 0 {
            return Err("sample rates must be positive");
        }
        Ok(Self {
            source_rate,
            canonical_rate,
        })
    }

    fn floor_to_source(self, sample: u64) -> u64 {
        let numerator = u128::from(sample).saturating_mul(u128::from(self.source_rate));
        u64::try_from(numerator / u128::from(self.canonical_rate)).unwrap_or(u64::MAX)
    }

    fn ceil_to_source(self, sample: u64) -> u64 {
        let denominator = u128::from(self.canonical_rate);
        let numerator = u128::from(sample)
            .saturating_mul(u128::from(self.source_rate))
            .saturating_add(denominator - 1);
        u64::try_from(numerator / denominator).unwrap_or(u64::MAX)
    }

    #[must_use]
    pub fn removal_to_source(self, interval: SampleInterval) -> SampleInterval {
        SampleInterval {
            start: self.floor_to_source(interval.start),
            end: self.ceil_to_source(interval.end),
        }
    }

    #[must_use]
    pub fn retained_to_source(self, interval: SampleInterval) -> SampleInterval {
        let start = self.ceil_to_source(interval.start);
        let end = self.floor_to_source(interval.end);
        SampleInterval {
            start,
            end: end.max(start),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn interval(start: u64, end: u64) -> SampleInterval {
        SampleInterval::new(start, end).expect("valid fixture")
    }

    #[test]
    fn half_open_intervals_do_not_share_boundary_samples() {
        let first = interval(0, 10);
        let second = interval(10, 20);
        assert!(!first.intersects(second));
        assert!(first.touches(second));
        assert!(!first.contains(10));
        assert!(second.contains(10));
    }

    #[test]
    fn normalize_merges_overlapping_and_abutting_intervals() {
        assert_eq!(
            normalize([
                interval(60, 70),
                interval(10, 20),
                interval(15, 30),
                interval(30, 40)
            ]),
            vec![interval(10, 40), interval(60, 70)]
        );
    }

    #[test]
    fn complement_returns_only_the_retained_gaps() {
        assert_eq!(
            complement(&[interval(10, 20), interval(50, 60)], interval(0, 100)),
            vec![interval(0, 10), interval(20, 50), interval(60, 100)]
        );
    }

    #[test]
    fn destination_map_uses_contiguous_output_offsets() {
        let map = DestinationMap::build([interval(0, 100), interval(300, 450)]);
        assert_eq!(map.destination_of(0), Some(interval(0, 100)));
        assert_eq!(map.destination_of(1), Some(interval(100, 250)));
        assert_eq!(map.map_sample(300), Some(100));
        assert_eq!(map.map_sample(200), None);
    }

    #[test]
    fn removal_rounds_outward_and_retention_rounds_inward() {
        let map = ResampleMap::new(44_100, 16_000).expect("valid rates");
        let removed = map.removal_to_source(interval(1, 2));
        let retained = map.retained_to_source(interval(1, 2));
        assert!(removed.start() <= retained.start());
        assert!(removed.end() >= retained.end());
    }
}
